{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "857d727f-ec2a-4819-8517-d06cd6233743",
   "metadata": {},
   "source": [
    "<center><font size = \"8\">Passive and Active membrane Properties<center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fabc5be-1d7c-4225-8ff2-a075d1251085",
   "metadata": {},
   "source": [
    "<font size=4><font color=blue> In this notebook you will learn:\n",
    "<ul style=\"color: blue; font-size: 18px;\">\n",
    "    <li> the differences between passive and active membrane properties and how this affects the signal propagation along the dendrites</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7edf02f8-0470-4ca5-87c2-ab6e08abbb0b",
   "metadata": {},
   "source": [
    "<p style=\"color: red; font-size: 16px;\"><b>IMPORTANT:</b> this notebook is using <b>Python</b> code to help you load, visualize and analyze the morphologies.</p>\n",
    "<ul style=\"color: red; font-size: 16px;\">\n",
    "  <li><b>It is not necessary to know how to code</b> (but it would be good for you if you try to understand the code)</li>\n",
    "  <li>To only run the cells: </li>\n",
    "    <ul style=\"color: red; font-size: 16px;\">\n",
    "        <li>Click once on one of the cells, you will notice that it is sellected because it is highlighted in blue.</li>\n",
    "        <li>Press \"Command+enter\" or click on the Run (triangle) button at the top of the page.</li>\n",
    "        <li>Consider that cells <b>must be run by order</b> </li>\n",
    "    </ul>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46869fc4-b9bb-4e3d-ba5e-0f601bb482bb",
   "metadata": {},
   "source": [
    "<p style=\"color: green; font-size: 16px;\"><b>YOUR GOAL:</b></p>\n",
    "<ul style=\"color: green; font-size: 16px;\">\n",
    "  <li>Answer the questions that you would find along the notebook.</li>\n",
    "  <li>For that you would have to understand what are the passive and the active properties of a neuron and how they affect signal propagation</li> \n",
    "  <li>  Build a report with your answeres and conclussions.</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44cc5bac-56b6-4d9a-af12-a0c5ef5dedb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell will load some useful files into your notebook environment\n",
    "\n",
    "import requests, os\n",
    "\n",
    "extra_files_root = \"https://raw.githubusercontent.com/NataliBZ/uc3m_notebooks/main/\"\n",
    "\n",
    "extra_files_names = [\n",
    "    \"Cellular/02_Passive_Active_properties/TTXDynamicsSwitch.mod\",\n",
    "    \"Cellular/02_Passive_Active_properties/UsefulFunctions.py\",\n",
    "    \"Cellular/02_Passive_Active_properties/active_neuron_1d_3d.py\",\n",
    "    \"Cellular/02_Passive_Active_properties/cell_01.swc\",\n",
    "    \"Cellular/02_Passive_Active_properties/cell_02.swc\",\n",
    "    \"Cellular/02_Passive_Active_properties/equivalent_circuit.png\",\n",
    "    \"Cellular/02_Passive_Active_properties/hoc2swc.py\",\n",
    "    \"Cellular/02_Passive_Active_properties/passive_neuron_1d_3d.py\",\n",
    "    \"Cellular/02_Passive_Active_properties/passive_neuron_1dend.py\"\n",
    "]\n",
    "for fn in extra_files_names:\n",
    "    req = requests.get(extra_files_root + fn)\n",
    "    if req.status_code == 200:\n",
    "        with open(os.path.split(fn)[1], \"wb\") as fid:\n",
    "            fid.write(req.content)\n",
    "    else:\n",
    "        print(\"Error obtaining auxiliary file!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "645be8a7-1e2a-4145-bae1-92a190c0828c",
   "metadata": {},
   "source": [
    "## 1. Passive and active membrane properties"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "403b3117-56b3-4b14-ae46-911deae30fae",
   "metadata": {},
   "source": [
    "<font size=\"4\">\n",
    "    \n",
    "__What are passive and active membrane properties?:__ \n",
    "\n",
    "- __Passive__ membrane properties are characteristics of a neuronal membrane, such as membrane resistance, membrane capacitance, and axial resistance, that do not depend on voltage-gated ion channels or receptors but are instead related to ion leakage channels, pumps, and the cell's physical structure. These properties influence how the membrane conducts electrical signals by affecting the efficiency of signal propagation, the speed of potential changes, and the integration of synaptic inputs. (You can know more about passive properties in the following [link](https://www.scientifica.uk.com/learning-zone/passive-membrane-properties))\n",
    "\n",
    "- __Active__ membrane properties depend on the activation of voltage gated ion channels. Unlike passive properties, active properties allow the neuron to generate and regulate action potentials, amplify or dampen inputs, and produce nonlinear responses.\n",
    "\n",
    "- This can be represented (modeled) as an RC circuit, with resistance and capacitance connected in parallel. Passive properties involve components that store, dissipate, or filter energy, such as resistors, capacitors, and inductors, and do not require an external power source to operate. For example, the lipid bylayer of the membrane acts as a capacitance with channels that allow the ions to pass through with certain resistance (membrane resistivity). In case of axons and dendrites, the intracelular axial resistance is a passive membrane property. This property leads to decay of the membrane potential in a distance dependent manner, starting from  the stimulation point. In contrast, active properties are associated with components that require an external power supply to amplify, generate, or control energy in the form of voltage or current, with examples including voltage/current sources and transistors. \n",
    "\n",
    "<center><img src=\"equivalent_circuit.png\" width=\"500\"><center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f3f250b-e9b2-4156-a67c-6db152fcc36e",
   "metadata": {},
   "source": [
    "### Let's start!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5b45428-9611-4629-bf66-897d56a79d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First call all the usefull packages\n",
    "%matplotlib inline\n",
    "\n",
    "from neuron import h\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import math as m\n",
    "import neurom as nm\n",
    "from neurom import view\n",
    "from hoc2swc import neuron2swc\n",
    "import UsefulFunctions as UF\n",
    "\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display, clear_output\n",
    "\n",
    "from passive_neuron_1dend import PassiveNeuron_1Dend\n",
    "from passive_neuron_1d_3d import PassiveNeuron_1D_3D\n",
    "from active_neuron_1d_3d import ActiveNeuron_1D_3D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9da36145-9ae3-4cde-a0c8-440abc6f266d",
   "metadata": {},
   "source": [
    "## 1.1 Signal attenuation due to passive membrane properties: \n",
    "## using a toy neuronal model: soma, 1 dendrite and 1 axon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97d19f61-cf29-494f-bb1d-ecbc58cb1dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_pass_1d = PassiveNeuron_1Dend()\n",
    "\n",
    "# Plot morphology so far\n",
    "UF.plot_morphology(\"cell_01\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f84c85d-5363-4790-aead-fe387bef4d1e",
   "metadata": {},
   "source": [
    "<font size=4> In this very simple and NON REALISTIC model:\n",
    "\n",
    "<ul style=\"color: black; font-size: 18px;\">\n",
    "    <li> the soma is represented as a black square of 20x20 µm</li>\n",
    "    <li> the axon (blue line) is defined by the following parameters: \n",
    "        <ul><li> diam = 3 µm </li></ul>\n",
    "        <ul><li> length = 100 µm </li></ul>\n",
    "        <ul><li> axial resistance = 123.0 Ω*cm </li></ul>\n",
    "        <ul><li> capacitance = 1 µF/cm^2 </li></ul>\n",
    "    <li> the dendrite (red line) is defined by the following parameters:</li>\n",
    "        <ul><li> diam = 1 µm </li></ul>\n",
    "        <ul><li> length = 200 µm </li></ul>\n",
    "        <ul><li> axial resistance = 300.0 Ω*cm </li></ul>\n",
    "        <ul><li> capacitance = 1 µF/cm^2 </li></ul>\n",
    "</ul>  \n",
    "\n",
    "<font size=3>*Note: if you are curious about how this neurons has been coded in python, check the file \"passive_neuron_1dend.py\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2140818-dcd2-44d1-98a3-75a01a1631fc",
   "metadata": {},
   "source": [
    "### Now we are going to see how the signal propagates into the soma from three different stimulation points in the dendrite."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4dfb258-4ee9-4166-8d79-d720f0ab5f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "UF.reset()\n",
    "\n",
    "# Scan at different locations with different delays\n",
    "delays = np.arange(100, 600, 200)\n",
    "locations = np.linspace(0, 1, 3)\n",
    "\n",
    "# Setup stimulations and recordings\n",
    "for p in zip(locations, delays):\n",
    "    UF.iclamp(cell_pass_1d.dend(p[0]), amplitude=0.4, delay=p[1], duration=50)\n",
    "UF.record_voltage(cell_pass_1d.soma(0.5))\n",
    "\n",
    "# Simulate and plot\n",
    "v_init = -70\n",
    "t_stop = 700\n",
    "UF.tvi_plots(UF.init_run(v_init, t_stop), UF.voltage_records, UF.current_records, vmax=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7db4a9-95df-4ec8-980f-4de2101db63f",
   "metadata": {},
   "source": [
    "<font size=4>You can see that the signal recording at the soma (voltages) is larger for the dendritic stimulation closer to it (dend(0)), while is smaller for the fardest dendritic stimulation (dend(1)). Just due to the physical properties of the neuron."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "637c1c76-889d-4895-8407-4d15fa9fb9ec",
   "metadata": {},
   "source": [
    "### Let's see what would happen if we stimulate at the soma and record at different points the dendrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "146946a0-a4f7-4497-9b87-6246f86d6f5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "UF.reset()\n",
    "\n",
    "# Scan at different locations with different delays\n",
    "locations = np.linspace(0, 1, 3)\n",
    "\n",
    "# Setup stimulations and recordings\n",
    "UF.iclamp(cell_pass_1d.soma(0.5), amplitude=0.4)\n",
    "for p in locations:\n",
    "    UF.record_voltage(cell_pass_1d.dend(p))\n",
    "\n",
    "# Simulate and plot\n",
    "v_init = -70\n",
    "t_stop = 700\n",
    "UF.tvi_plots(UF.init_run(v_init, t_stop), UF.voltage_records, UF.current_records, vmax=-10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36699ffa-7fdf-42fc-88e6-437fd625cb4a",
   "metadata": {},
   "source": [
    "## INTERACTIVE:\n",
    "### Play around with the following passive parameters and observe how the voltage changes along the dendrite:\n",
    "    - dendritic diamater\n",
    "    - dendritic axial resistance\n",
    "    - dendritic capacitance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "586b3b09-66b1-4a86-ad80-fd09fabc2742",
   "metadata": {},
   "outputs": [],
   "source": [
    "UF.chage_passive_prop(cell_pass_1d)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f2c2977-cf61-4140-bbca-9a08783341e6",
   "metadata": {},
   "source": [
    "<font size=6><font color=blue> Question 1:\n",
    "\n",
    "<font size=4><font color=blue>  After exploring different values for the passive properties such as dendritic diameter, axial resistance and dendritic capacitance, explain what is the relationship between the signal propagation and each of the explored passive properties.\n",
    "\n",
    "<font size=4><font color=blue>As observed, stimulating the dendrite at points closer to the soma (e.g., dend(0)) results in a larger voltage signal recorded at the soma than stimulation from points farther away (e.g., dend(1)). Explain the biophysical properties of the neuron that account for this decrease in signal strength with increasing distance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84dda84a-9a6c-4021-beec-be43af7ae226",
   "metadata": {},
   "source": [
    "## 1.2 Signal attenuation due to passive membrane properties: \n",
    "### using a toy neuronal model: soma, dendrite with 4 dendritic branches with different diametres and 1 axon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a393834-2cb5-4d2f-a3bd-b020a1054089",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_pass_1d_3d = PassiveNeuron_1D_3D()\n",
    "\n",
    "# Plot morphology so far\n",
    "UF.plot_morphology(\"cell_02\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "867292ef-be09-4b63-8c6d-5d8991f11c45",
   "metadata": {},
   "source": [
    "### As we did before we will stimulate at the final end of each dendrite and record the voltage signal at the soma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97d1a3cc-9dfc-4d4b-a708-674523043ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "UF.reset()\n",
    "\n",
    "branchlst = [cell_pass_1d_3d.branch_0, cell_pass_1d_3d.branch_1, cell_pass_1d_3d.branch_2, cell_pass_1d_3d.branch_3]\n",
    "\n",
    "# Setup stimulations and recordings\n",
    "for idx, branch in enumerate(branchlst):\n",
    "    UF.iclamp(branch(0.5), amplitude=0.4, delay=200 + idx * 100, duration=50)\n",
    "UF.record_voltage(cell_pass_1d_3d.soma(0.5))\n",
    "\n",
    "# Simulate and plot\n",
    "v_init = -70\n",
    "t_stop = 700\n",
    "UF.tvi_plots(UF.init_run(v_init, t_stop), UF.voltage_records, UF.current_records, vmax=-65)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83c22335-4fc6-47e9-8b94-fc3137e2e2bf",
   "metadata": {},
   "source": [
    "### As before we will stimulate at the soma and record at the end of each dendrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24160f59-c432-4b58-9868-877dfd84e85d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "UF.reset()\n",
    "\n",
    "# Setup stimulations and recordings\n",
    "UF.iclamp(cell_pass_1d_3d.soma(0.5), amplitude=0.4)\n",
    "for branch in branchlst:\n",
    "    UF.record_voltage(branch(1))\n",
    "\n",
    "# Simulate and plot\n",
    "v_init = -70\n",
    "t_stop = 700\n",
    "UF.tvi_plots(UF.init_run(v_init, t_stop), UF.voltage_records, UF.current_records, vmax=-66)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2c2a037-7ae6-4a34-bede-ce358079abe2",
   "metadata": {},
   "source": [
    "<font size=6><font color=blue> Question 2:\n",
    "\n",
    "<font size=4><font color=blue>  Based on your answer to question 1 regarding dendritic diameter, how would you explain what is happening to the signal propagation into the soma or through the dendritic branches in the previous example of point 1.2\n",
    "\n",
    "<font size=4><font color=blue>Based on the parameters modeled from both neuronal structures built, why is the axial resistance higher in the dendrite than in the axon? Which specific biophysical factors are behind this difference? (Hint: check how the models are built in files \"passive_neuron_1dend.py\" and \"passive_neuron_1d_3d.py\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2229b180-c772-43bf-a2fe-dba8151b2be8",
   "metadata": {},
   "source": [
    "## 1.3 Active properties.  \n",
    "## We will perform the same experiments but including active properties at the soma and the dendrites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c90b319b-6a4d-491f-930f-5074f87fd6d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Firts we load the neuron with active properties:\n",
    "\n",
    "cell_active_1d_3d = ActiveNeuron_1D_3D()\n",
    "\n",
    "# Plot morphology so far\n",
    "UF.plot_morphology(\"cell_02\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26aa076f-d4b9-4629-bc89-b98095a21b70",
   "metadata": {},
   "source": [
    "# Back-propagating action potential.\n",
    "<p style=\"font-size: 16px;\">\n",
    "Normally, when a neuron fires an action potential, it starts at the axon hillock (near the cell body) and travels forward along the axon to send signals to other cells, that is called forward propagation.\n",
    "</p>\n",
    "<p style=\"font-size: 16px;\">\n",
    "At the same time, a copy of that electrical signal also travels backward, from the soma (cell body) into the dendrites. That is called a back-propagating action potential (bAP).\n",
    "</p>\n",
    "\n",
    "<p style=\"font-size: 16px;\"><b>Why does it happen?</b></p>\n",
    "<p style=\"font-size: 16px;\">\n",
    "Because dendrites are made of excitable membrane too, they have voltage-gated sodium and potassium channels, just like the axon.\n",
    "So when the action potential starts, some of that depolarization can spread backward and briefly activate those channels in the dendrites.\n",
    "</p>\n",
    "\n",
    "<p style=\"font-size: 16px;\"><b>Does it has a functionality?</b></p>\n",
    "\n",
    "The bAP does not send a message to other neurons, but it’s very important inside the neuron itself:\n",
    "<ul style=\"font-size: 16px;\">\n",
    "        <li>It tells the dendrites that the cell has fired — kind of a “heads-up” signal.</li>\n",
    "        <li>It helps in synaptic plasticity (learning): when a bAP coincides with input at a synapse, that connection can get stronger or weaker (this is part of spike- timing–dependent plasticity).</li>\n",
    "        <li>It can help regulate ion channel distribution and gene expression in dendrites.</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eca881b-b400-4c47-a67c-d08a28a8b8ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "UF.reset()\n",
    "\n",
    "branchlst = [cell_active_1d_3d.branch_0, cell_active_1d_3d.branch_1, cell_active_1d_3d.branch_2, cell_active_1d_3d.branch_3]\n",
    "\n",
    "# Setup stimulations and recordings\n",
    "for branch in branchlst:\n",
    "    UF.iclamp(branch(1), amplitude=0.5, delay=100, duration=50)\n",
    "\n",
    "UF.record_voltage(cell_active_1d_3d.soma(0.5))\n",
    "UF.record_voltage(cell_active_1d_3d.dend(0.5))\n",
    "\n",
    "# Simulate and plot\n",
    "v_init = -70\n",
    "t_stop = 700\n",
    "UF.tvi_plots(UF.init_run(v_init, t_stop), UF.voltage_records, UF.current_records, vmax=60, imax=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8727690a-cfd5-458b-b0dc-e1b5abeddbe5",
   "metadata": {},
   "source": [
    "<font size=6><font color=blue> Question 3:\n",
    "\n",
    "<font size=4><font color=blue>  Imagine that you are studying how the signal propagates along two different neurons, one with dendrites two times ticker than the other. Considering only the passive properties, use what you have learnt during the lecture and this practical session to hypothesised which neuron would be easier to activate under the same stimulation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1caa6a7a-f8a8-4a27-b0e1-adbe558341e4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<font size=6><font color=blue> Question 4:\n",
    "\n",
    "<font size=4><font color=blue>  Imagine now that the neuron with thicker dendrites also has four times higher active membrane conductance (i.e., due to ion channels), which neuron would be easier to activate under the same stimulation, and why?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "427ce550-9320-42a2-90de-71d629e6f0cf",
   "metadata": {},
   "source": [
    "<ul style=\"color: green; font-size: 24px;\">\n",
    "  If this notebook sparked your interest about studying single cell behavior, feel free to explore the data and build your own single cell model on the Open Brain Platform.\n",
    "</ul>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
